src/coreclr/vm/jithelpers.cpp

```diff
@@ -3618,3 +3426,212 @@ void InitJITHelpers2()
     SetJitHelperFunction(CORINFO_HELP_INIT_PINVOKE_FRAME, (void *)GenerateInitPInvokeFrameHelper()->GetEntryPoint());
 #endif // TARGET_X86 || TARGET_ARM
 }
+
+
+#if defined(TARGET_32BIT) && !defined(TARGET_X86)
+
+// Using the musttail attribute requires that we disble our protection against using return in the wrong place
+#ifdef return
+#undef return
+#endif
+
+#if TARGET_UNIX
+#ifdef __clang__
+#define MUSTTAIL [[clang::musttail]]
```

Comment:
BTW, the performance concern was only for win-x86 and performance optimizations aren't critical for linux-arm and linux-x86. Moreover, it seems like MUSTTAIL is only to accelerate exception path with no effect on the actual div/mod ops? If this is the case, then I think we can just drop this optimization even if it was for win-x86 because perf concern was about the div/mod (non-exception) happy path.

Reply:
The MUSTTAIL logic here isn't an optimization as without HMFs we can't through an exception from an FCALL, but I agree, it doesn't look like it has worked out as a reasonable way to make this work simple. My current plan is to keep the work I've done to port the Windows X86 helpers to handcoded assembly, and merge it with @am11's work that is more general purpose, and then call it good.